File: JSBridge/Classes/JSCore/XMLHTTPRequest.m
/*
 * Copyright (C) 2019 ProSiebenSat1.Digital GmbH.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "XMLHTTPRequest.h"

static NSURLSession *_urlSession;
static dispatch_queue_t _jsQueue;
static NSPointerArray *_instances = nil;

@implementation XMLHttpRequest {
    NSString *_httpMethod;
    NSURL *_url;
    bool _async;
    bool _isAborted;
    NSMutableDictionary *_requestHeaders;
    NSDictionary *_responseHeaders;
    NSString *_responseType;
};

@synthesize response;
@synthesize responseText;
@synthesize responseType;
@synthesize onreadystatechange;
@synthesize readyState;
@synthesize onload;
@synthesize onabort;
@synthesize onprogress;  // TODO
@synthesize onerror;     // TODO
@synthesize status;
@synthesize statusText;
@synthesize withCredentials;
@synthesize onCompleteHandler;
@synthesize loggingHandler;

+ (void)globalInit {
    [XMLHttpRequest globalInitWithURLSession:[NSURLSession sharedSession] jsQueue:dispatch_get_main_queue()];
}

+ (void)globalInitWithURLSession:(NSURLSession *)urlSession {
    [XMLHttpRequest globalInitWithURLSession:urlSession jsQueue:dispatch_get_main_queue()];
}

+ (void)globalInitWithURLSession:(NSURLSession *)urlSession jsQueue:(dispatch_queue_t)jsQueue {
    _urlSession = urlSession;
    _jsQueue = jsQueue;
}

- init {
    if (self = [super init]) {
        self.readyState = @(XMLHttpRequestUNSENT);
        self.responseType = @"";

        _requestHeaders = [NSMutableDictionary new];

        // Set a default user agent in the form "Apple iPhone iOS 10.0"
        UIDevice *currentDevice = [UIDevice currentDevice];
        NSString *userAgent = [NSString stringWithFormat:@"Apple %@ %@ %@",
                               [currentDevice model], [currentDevice systemName], [currentDevice systemVersion]];

        [_requestHeaders setValue:userAgent forKey:@"User-Agent"];
    }
    return self;
}

+ (void)extend:(id)jsContext onNewInstance:(void(^)(XMLHttpRequest *))onNewInstance {
    dispatch_async(_jsQueue, ^{
        // Simulate the constructor
        jsContext[@"XMLHttpRequest"] = ^{
            XMLHttpRequest *instance = [XMLHttpRequest new];
            onNewInstance(instance);
            return instance;
        };
        jsContext[@"XMLHttpRequest"][@"UNSENT"] = @(XMLHttpRequestUNSENT);
        jsContext[@"XMLHttpRequest"][@"OPENED"] = @(XMLHttpRequestOPENED);
        jsContext[@"XMLHttpRequest"][@"LOADING"] = @(XMLHttpRequestLOADING);
        jsContext[@"XMLHttpRequest"][@"HEADERS"] = @(XMLHttpRequestHEADERS);
        jsContext[@"XMLHttpRequest"][@"DONE"] = @(XMLHttpRequestDONE);
    });
}

- (void)clearJSValues {
    // Null all JSValue instances to avoid a retain cycle causing the JSContext instance to be leaked
    self.onreadystatechange = nil;
    self.onload = nil;
    self.onabort = nil;
    self.onprogress = nil;
    self.onerror = nil;
}

- (void)open:(NSString *)httpMethod :(NSString *)url :(bool)async {
    // TODO should throw an error if called with wrong arguments
    _httpMethod = httpMethod;
    _url = [NSURL URLWithString:url];
    _async = async;
    self.readyState = @(XMLHttpRequestOPENED);
}

- (void)send:(id)data {
    NSURL *url = _url;

    // handle invalid URLs (often no scheme or invalid characters like curly brackets
    // can cause NSURL object not be created in open:
    if (url == nil) {
        [self.onerror callWithArguments:@[]];
        return;
    }

    self.readyState = @(XMLHttpRequestLOADING);
    [self.onreadystatechange callWithArguments:@[]];
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    for (NSString *name in _requestHeaders) {
        [request setValue:_requestHeaders[name] forHTTPHeaderField:name];
    }
    if ([data isKindOfClass:[NSString class]]) {
        request.HTTPBody = [((NSString *) data) dataUsingEncoding:NSUTF8StringEncoding];
    }
    [request setHTTPMethod:_httpMethod];

    __block __weak XMLHttpRequest *weakSelf = self;
    id completionHandler = ^(NSData *data, NSURLResponse *response, NSError *error) {
        XMLHttpRequest *strongSelf = weakSelf;
        if (strongSelf == nil) {
            return;
        }

        dispatch_async(_jsQueue, ^{
            [strongSelf handleResponse:response data:data error:error];
        });
    };

    NSURLSessionDataTask *task = [_urlSession dataTaskWithRequest:request completionHandler:completionHandler];
    [task resume];
}

- (void)handleResponse:(NSURLResponse*)receivedResponse data:(NSData*)data error:(NSError*)error {
    if ([self.readyState isEqual:@(XMLHttpRequestUNSENT)]) {
        [self.onabort callWithArguments:@[]];
        return;
    } else if (![self.readyState isEqual:@(XMLHttpRequestLOADING)]) {
        return;
    }

    if (error != nil) {
        [self.onerror callWithArguments:@[]];
        return;
    }

    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)receivedResponse;
    readyState = @(XMLHttpRequestDONE);
    status = @(httpResponse.statusCode);
    statusText = [NSString stringWithFormat:@"%lid", (long)httpResponse.statusCode];
    responseText = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];

    [self setAllResponseHeaders:httpResponse.allHeaderFields];

    self.response = [self getResponseWithResponseType:responseType responseText:responseText];

    [onreadystatechange callWithArguments:@[]];
    [onload callWithArguments:@[]];

    if (loggingHandler) {
        NSString *message = [NSString stringWithFormat:@"[JSBridge] request: %@ response: %ld %@",
                             _url,
                             (long)httpResponse.statusCode,
                             responseText];
        loggingHandler(message);
    }

    // Make sure that the XMLHttpRequest instance can be deallocated by nulling the JSValue instances
    // (which retain the JSContext)
    [self clearJSValues];

    // call onCompleteHandler, required to support Promise on older JSCore versions
    if (onCompleteHandler) onCompleteHandler();

}

- (void)abort {
    // Note: according to the XHR specs, onreadystate() is not supposed to be triggered
    self.readyState = @(XMLHttpRequestUNSENT);
}

- (void)setRequestHeader:(NSString *)name :(NSString *)value {
    _requestHeaders[name] = value;
}

- (NSString *)getAllResponseHeaders {
    NSMutableString *responseHeaders = [NSMutableString new];
    for (NSString *key in _responseHeaders) {
        [responseHeaders appendString:key];
        [responseHeaders appendString:@": "];
        [responseHeaders appendString:_responseHeaders[key]];
        [responseHeaders appendString:@"\r\n"];
    }
    return responseHeaders;
}

- (NSString *)getResponseHeader:(NSString *)name {
    return _responseHeaders[[name lowercaseString]];
}
- (void)setAllResponseHeaders:(NSDictionary *)responseHeaders {
    // Convert to lower case for case-insentive lookup
    NSMutableDictionary *dict = [NSMutableDictionary new];
    [responseHeaders enumerateKeysAndObjectsUsingBlock:^(id key, id value, BOOL *stop) {
        NSString *lowerCaseKey = [(NSString *)key lowercaseString];
        dict[lowerCaseKey] = value;
    }];

    _responseHeaders = dict;
}

- (id)getResponseWithResponseType:(NSString * _Nonnull)contentType responseText:(NSString * _Nonnull)responseText {
    if ([contentType isEqualToString:@""]) {
        return responseText;
    }
    
    if ([contentType isEqualToString:@"text"]) {
        return responseText;
    }

    if ([contentType isEqualToString:@"json"]) {
        NSData *data = [responseText dataUsingEncoding:NSUTF8StringEncoding];
        return [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
    }

    // TODO: support more response types like "arraybuffer", "document"
    return responseText;
}

@end

File: JSBridge/Classes/JSCore/XMLHttpRequest.swift
import Foundation
import JavaScriptCore

/// https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest
/// https://xhr.spec.whatwg.org/#interface-xmlhttprequest
@objc protocol XMLHttpRequestJSExport: JSExport {
    // Instance properties

    var readyState: NSNumber { get set }
    var response: Any? { get set }
    var responseText: String? { get set }
    var responseType: String { get set }
    var status: NSNumber { get set }

    // Events

    // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#events
    var onreadystatechange: JSValue? { get set }
    var onload: JSValue? { get set }
    var onsend: JSValue? { get set }
    var onabort: JSValue? { get set }
    var onerror: JSValue? { get set }

    /// https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener
    func addEventListener(_ type: String, _ handler: JSValue)
    /// https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener
    func removeEventListener(_ type: String, _ handler: JSValue)

    // Instance methods

    func open(_ httpMethod: String, _ url: String)
    func send(_ data: Any?)
    func abort()
    func setRequestHeader(_ header: String, _ value: String)
    func getAllResponseHeaders() -> String
    func getResponseHeader(_ name: String) -> String?
}

/**
 Native implementation of XMLHttpRequest object to request data from a server.
 It is available out-of-the-box in all modern browsers and in WebKit, but not available in JavaScriptCore.

 Limitations of the native implementation:
 - Only a limited set of event callback properties is available, but all event listener events are available.
 - Only the first event listener registered for each event type will be called.
 - Requests cannot be reused. To avoid capturing `JSContext`, all request callbacks and listeners are removed after handling the response.
 */
@objc class XMLHttpRequest: NSObject {
    /// https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
    enum ReadyState: NSNumber {
        case unsent = 0, opened, headersReceived, loading, done
    }

    // MARK: Internal properties

    dynamic var readyState = ReadyState.unsent.rawValue
    dynamic var response: Any?
    dynamic var responseText: String?
    dynamic var responseType = ""
    dynamic var status: NSNumber = 0

    dynamic var onreadystatechange: JSValue?
    dynamic var onload: JSValue?
    dynamic var onsend: JSValue?
    dynamic var onabort: JSValue?
    dynamic var onerror: JSValue?

    // MARK: Private properties

    private let jsQueue: DispatchQueue
    private let urlSession: URLSession
    private let logger: ((String) -> Void)?
    private var eventListeners = [XMLHttpRequestEvent.EventType: JSValue]()

    private var request: URLRequest?
    private var dataTask: URLSessionDataTask?
    private var responseHeaders = [String: String]()
    private var responseHeadersString = ""

    // MARK: Init

    init(urlSession: URLSession, jsQueue: DispatchQueue, logger: ((String) -> Void)?) {
        self.urlSession = urlSession
        self.jsQueue = jsQueue
        self.logger = logger

        super.init()
    }

    // MARK: Internal static methods

    static func configure(
        urlSession: URLSession,
        jsQueue: DispatchQueue,
        context: JSContext,
        logger: ((String) -> Void)? = nil
    ) {
        jsQueue.async {
            let constructor: @convention(block) () -> XMLHttpRequest = {
                XMLHttpRequest(urlSession: urlSession, jsQueue: jsQueue, logger: logger)
            }
            context.setObject(constructor, forKeyedSubscript: NSString(string: "XMLHttpRequest"))

            let xmlRequest = context.objectForKeyedSubscript("XMLHttpRequest")!
            xmlRequest.setObject(ReadyState.unsent.rawValue, forKeyedSubscript: NSString(string: "UNSENT"))
            xmlRequest.setObject(ReadyState.opened.rawValue, forKeyedSubscript: NSString(string: "OPENED"))
            xmlRequest.setObject(ReadyState.loading.rawValue, forKeyedSubscript: NSString(string: "LOADING"))
            xmlRequest.setObject(ReadyState.headersReceived.rawValue, forKeyedSubscript: NSString(string: "HEADERS_RECEIVED"))
            xmlRequest.setObject(ReadyState.done.rawValue, forKeyedSubscript: NSString(string: "DONE"))
        }
    }
}

// MARK: - JSExport implementation

extension XMLHttpRequest: XMLHttpRequestJSExport {
    func open(_ httpMethod: String, _ urlString: String) {
        if let url = URL(string: urlString.trimmingCharacters(in: .whitespaces)) {
            var request = URLRequest(url: url)
            request.httpMethod = httpMethod
            self.request = request
        } else {
            log("Cannot create URL from \(urlString)")
            self.request = nil
        }

        readyState = ReadyState.opened.rawValue
        emitEvent(type: .readyStateChange)
    }

    func send(_ data: Any?) {
        // DOMException: Failed to execute 'send' on 'XMLHttpRequest': The object's state must be OPENED.
        guard readyState == ReadyState.opened.rawValue else {
            log("Failed to execute 'send' on 'XMLHttpRequest': The object's state must be \(ReadyState.opened.rawValue) (OPENED), but it is \(readyState)")
            return
        }

        emitEvent(type: .loadStart)

        // Requests created with invalid URL shall fail.
        guard var request else {
            finishWithError()
            return
        }

        if let payload = data as? String {
            request.httpBody = payload.data(using: .utf8)
        }

        let dataTask = urlSession.dataTask(with: request) { [weak self] data, response, error in
            self?.jsQueue.async {
                self?.processResponse(data, response, error)
            }
        }
        self.dataTask = dataTask
        dataTask.resume()
    }

    func setRequestHeader(_ header: String, _ value: String) {
        request?.setValue(value, forHTTPHeaderField: header)
    }

    func abort() {
        if let dataTask {
            dataTask.cancel()
            self.dataTask = nil

            readyState = ReadyState.done.rawValue
            emitEvent(type: .readyStateChange)

            emitEvent(type: .abort)
            emitEvent(type: .loadEnd)
        }

        readyState = ReadyState.unsent.rawValue
        status = 0
    }

    func getAllResponseHeaders() -> String {
        responseHeadersString
    }

    func getResponseHeader(_ name: String) -> String? {
        responseHeaders[name.lowercased()]
    }

    func addEventListener(_ type: String, _ handler: JSValue) {
        guard let eventType = XMLHttpRequestEvent.EventType(rawValue: type), eventListeners[eventType] == nil else { return }

        eventListeners[eventType] = handler
    }

    func removeEventListener(_ type: String, _ handler: JSValue) {
        guard let eventType = XMLHttpRequestEvent.EventType(rawValue: type) else { return }

        eventListeners[eventType] = nil
    }

    // MARK: Private methods

    private func log(_ message: String) {
        logger?(message)
    }

    /// Links XHR Event Type to property callbacks
    private func eventProperty(eventType: XMLHttpRequestEvent.EventType) -> JSValue? {
        switch eventType {
        case .abort:
            return onabort
        case .error:
            return onerror
        case .load:
            return onload
        case .loadStart:
            return onsend
        case .readyStateChange:
            return onreadystatechange
        default:
            return nil
        }
    }

    private func emitEvent(type: XMLHttpRequestEvent.EventType) {
        let property = eventProperty(eventType: type)
        let listener = eventListeners[type]

        // Skip creating event payload if there are no callbacks for given eventType
        guard property != nil || listener != nil else { return }

        let eventPayload = XMLHttpRequestEvent(type: type, value: self)
        property?.call(withArguments: [eventPayload])
        listener?.call(withArguments: [eventPayload])
    }

    /// Nullifies all JSValue instances to avoid a retain cycle causing the JSContext instance to be leaked
    func clearJSValues() {
        onreadystatechange = nil
        onload = nil
        onsend = nil
        onabort = nil
        onerror = nil

        eventListeners.removeAll()
    }

    /// Performs a chain of actions which happens when XHR
    ///   - has been opened and sent with invalid/broken URL;
    ///   - fails with a network error.
    private func finishWithError() {
        readyState = ReadyState.done.rawValue
        emitEvent(type: .readyStateChange)

        emitEvent(type: .error)
        emitEvent(type: .loadEnd)
    }

    private func processResponse(_ data: Data?, _ response: URLResponse?, _ error: Error?) {
        defer { clearJSValues() }

        if let error = error as? NSError, error.domain == NSURLErrorDomain, error.code == NSURLErrorCancelled {
            return
        }

        guard let response = response as? HTTPURLResponse, error == nil else {
            finishWithError()
            return
        }

        if !(200 ..< 300 ~= response.statusCode) {
            log("XHR response returned with status code \(response.statusCode) for \"\(String(describing: response.url))\"")
        }

        status = NSNumber(value: response.statusCode)
        decodeResponse(data: data, responseType: responseType)

        for field in response.allHeaderFields {
            guard
                let key = (field.key as? String)?.lowercased(),
                let value = field.value as? String else { continue }
            responseHeadersString += (key + ": " + value + "\r\n")
            responseHeaders[key] = value
        }

        readyState = ReadyState.headersReceived.rawValue
        emitEvent(type: .readyStateChange)

        readyState = ReadyState.loading.rawValue
        emitEvent(type: .readyStateChange)

        emitEvent(type: .progress)

        readyState = ReadyState.done.rawValue
        emitEvent(type: .readyStateChange)

        emitEvent(type: .load)
        emitEvent(type: .loadEnd)
    }

    private func decodeResponse(data: Data?, responseType: String) {
        // Reset fields state
        responseText = nil
        response = nil

        guard let data else { return }

        switch responseType {
        case "json":
            response = try? JSONSerialization.jsonObject(with: data)
        default:
            responseText = String(data: data, encoding: .utf8)
            response = responseText
        }
    }
}
